File: web/src/api/evaluations.ts
import client from './client'
import type { Employee } from './employees'

export interface EvaluationRecord {
  id: number
  employee_id: number
  employee_name: string | null
  target_level: string | null
  reviewer_name: string
  status: string
  scores: (number | null)[]
  avg_values: number | null
  avg_capability: number | null
  avg_output: number | null
  final_score: number | null
  sys_suggestion: string | null
  reviewer_result: string | null
  advantage: string | null
  disadvantage: string | null
  create_time: string
  update_time: string
}

export interface LoadEvaluationResponse {
  employee: Employee
  record: EvaluationRecord | null
}

export interface DraftPayload {
  employee_id: number
  reviewer_name: string
  scores: (number | null)[]
  advantage?: string | null
  disadvantage?: string | null
}

export interface GeneratePayload extends DraftPayload {
  sys_suggestion: string
  reviewer_result: string
}

export async function loadEvaluation(
  employeeId: number,
  reviewerName: string,
): Promise<LoadEvaluationResponse> {
  const res = await client.get('/evaluations/load', {
    params: { employee_id: employeeId, reviewer_name: reviewerName },
  })
  return res.data
}

export async function saveDraft(data: DraftPayload): Promise<EvaluationRecord> {
  const res = await client.post('/evaluations/draft', data)
  return res.data
}

export async function generateResult(data: GeneratePayload): Promise<EvaluationRecord> {
  const res = await client.post('/evaluations/generate', data)
  return res.data
}

export async function submitEvaluation(recordId: number): Promise<EvaluationRecord> {
  const res = await client.post('/evaluations/submit', { record_id: recordId })
  return res.data
}

File: backend/app/routers/evaluations.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.auth import require_admin
from app.database import get_db
from app.models import UserInfo
from app.response import fail, ok
from app.schemas import (
    EvaluationDraftRequest,
    EvaluationGenerateRequest,
    EvaluationSubmitRequest,
    UserInfoResponse,
)
from app.services.evaluation import (
    generate_result,
    load_evaluation,
    query_summary,
    record_to_dict,
    submit_record,
    upsert_draft,
)
from app.services.excel import build_summary_export

router = APIRouter()


@router.get("/load")
def load_eval(
    employee_id: int = Query(...),
    reviewer_name: str = Query(...),
    db: Session = Depends(get_db),
):
    if not reviewer_name.strip():
        return fail("评委姓名不能为空")
    emp = db.get(UserInfo, employee_id)
    if not emp:
        return fail("员工不存在", status_code=404)

    rec = load_evaluation(db, employee_id, reviewer_name.strip())
    data = {
        "employee": UserInfoResponse.model_validate(emp).model_dump(),
        "record": record_to_dict(rec, emp) if rec else None,
    }
    return ok(data)


@router.post("/draft")
def save_draft(body: EvaluationDraftRequest, db: Session = Depends(get_db)):
    try:
        rec = upsert_draft(
            db, body.employee_id, body.reviewer_name.strip(),
            body.scores, body.advantage, body.disadvantage,
        )
        emp = db.get(UserInfo, body.employee_id)
        return ok(record_to_dict(rec, emp))
    except ValueError as e:
        return fail(str(e), status_code=409)


@router.post("/generate")
def generate(body: EvaluationGenerateRequest, db: Session = Depends(get_db)):
    if len(body.scores) != 12 or any(s is None for s in body.scores):
        return fail("12项分数必须全部填写")
    if not body.advantage or not body.disadvantage:
        return fail("突出优势和待发展项必填")
    try:
        rec = generate_result(
            db, body.employee_id, body.reviewer_name.strip(),
            body.scores, body.advantage, body.disadvantage,
            body.sys_suggestion, body.reviewer_result,
        )
        emp = db.get(UserInfo, body.employee_id)
        return ok(record_to_dict(rec, emp))
    except ValueError as e:
        return fail(str(e), status_code=409)


@router.post("/submit")
def submit(body: EvaluationSubmitRequest, db: Session = Depends(get_db)):
    try:
        rec = submit_record(db, body.record_id)
        emp = db.get(UserInfo, rec.employee_id)
        return ok(record_to_dict(rec, emp))
    except ValueError as e:
        return fail(str(e), status_code=409)


@router.get("/summary")
def summary(
    employee_name: Optional[str] = Query(None),
    reviewer_name: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    _: str = Depends(require_admin),
):
    return ok(query_summary(db, employee_name, reviewer_name))


@router.get("/export")
def export(
    employee_name: Optional[str] = Query(None),
    reviewer_name: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    _: str = Depends(require_admin),
):
    rows = query_summary(db, employee_name, reviewer_name)
    content = build_summary_export(rows)
    return Response(
        content=content,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": "attachment; filename=evaluation_summary.xlsx"},
    )
